File: modules/08-classification/py/04_logistic_scratch.py
"""
Task 4 🔴 — Multinomial logistic regression FROM SCRATCH.

What you'll learn:
  - The softmax function and why it turns raw scores into probabilities
  - Cross-entropy loss: the standard loss for classification
  - Gradient descent: how the weight matrix updates to minimise loss
  - Why the gradient of softmax + cross-entropy is elegantly simple

The math (README explains each step in plain English):

  Forward pass:
    z = X @ W + b           (linear layer: batch of scores, shape [N, C])
    p = softmax(z)          (probabilities, each row sums to 1)
    L = cross_entropy(p, y) (scalar loss)

  Backward pass (gradient of L with respect to W and b):
    dL/dz = p - one_hot(y)  (elegant: error is just predicted - true probs)
    dL/dW = X.T @ dL/dz / N
    dL/db = mean(dL/dz, axis=0)

  Update:
    W -= lr * dL/dW
    b -= lr * dL/db

No sklearn in this task. Only numpy.

How to run:
  uv run python modules/08-classification/py/04_logistic_scratch.py

The harness loads the dataset, embeds the texts, trains the model, and prints
training loss + final test accuracy. The gradient step is left as a TODO so you
implement the learning yourself.
"""

from __future__ import annotations

import json
import pathlib

import numpy as np
from llm_core import get_provider

DATA_PATH = pathlib.Path(__file__).parent.parent / "data" / "texts.json"
LABEL_NAMES = ["technology", "science", "business", "sports", "health", "politics"]
NUM_CLASSES = len(LABEL_NAMES)

# ---------------------------------------------------------------------------
# Data loading and embedding
# ---------------------------------------------------------------------------


def load_and_embed(provider) -> tuple[np.ndarray, np.ndarray]:
    """
    Load the dataset and embed all texts.

    Returns:
      X : float32 array of shape (N, D) — embedding matrix
      y : int array of shape (N,)     — label indices
    """
    with open(DATA_PATH) as f:
        data = json.load(f)

    texts = [item["text"] for item in data]
    label_to_idx = {l: i for i, l in enumerate(LABEL_NAMES)}
    y = np.array([label_to_idx[item["label"]] for item in data], dtype=np.int64)

    # Embed in one batch (50 texts is well within limits)
    result = provider.embed(texts)
    X = np.array(result.vectors, dtype=np.float32)
    return X, y


# ---------------------------------------------------------------------------
# Train/test split (same stratified 80/20 as other tasks)
# ---------------------------------------------------------------------------


def stratified_split(
    X: np.ndarray, y: np.ndarray, test_fraction: float = 0.2, seed: int = 42
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Stratified train/test split. Returns (X_train, X_test, y_train, y_test)."""
    rng = np.random.default_rng(seed)
    train_idx, test_idx = [], []
    for cls in np.unique(y):
        idx = np.where(y == cls)[0]
        rng.shuffle(idx)
        n_test = max(1, int(len(idx) * test_fraction))
        test_idx.extend(idx[:n_test].tolist())
        train_idx.extend(idx[n_test:].tolist())
    return X[train_idx], X[test_idx], y[train_idx], y[test_idx]


# ---------------------------------------------------------------------------
# Softmax and cross-entropy — implement these
# ---------------------------------------------------------------------------


def softmax(z: np.ndarray) -> np.ndarray:
    """
    Softmax over the last axis.

    softmax(z)_i = exp(z_i) / sum_j(exp(z_j))

    Numerically stable version: subtract max(z) before exp() to prevent overflow.
    This doesn't change the result because exp(z - max) / sum(exp(z - max))
    = exp(z) / sum(exp(z)).

    Shape: z is (N, C) → returns (N, C), each row sums to 1.

    TODO: implement the numerically-stable softmax.
      - First shift each row so its max becomes 0: subtract the per-row max.
        Use `np.max(..., axis=1, keepdims=True)` so it broadcasts back over rows.
      - Exponentiate the shifted scores, then divide each row by its own sum
        (again reduce with `axis=1, keepdims=True` so the shapes broadcast).
    """
    # TODO: implement numerically-stable softmax
    raise NotImplementedError("TODO: implement softmax()")


def cross_entropy_loss(probs: np.ndarray, y: np.ndarray) -> float:
    """
    Mean cross-entropy loss over a batch.

    L = -1/N * sum_i( log( p[i, y[i]] ) )

    p[i, y[i]] is the predicted probability for the TRUE class of sample i.
    We want this to be 1.0 (log 1 = 0, no loss); if it's tiny, log is very negative.

    Clip probabilities at 1e-12 to avoid log(0).

    TODO: implement.
      - Clip `probs` into the range [1e-12, 1.0] with `np.clip` before taking any log.
      - Pull out the probability assigned to each sample's TRUE class. Advanced
        indexing does this in one shot: index rows with `np.arange(N)` and columns
        with `y`.
      - The loss is the negative mean of the logs of those correct-class probs.
    """
    # TODO: implement
    raise NotImplementedError("TODO: implement cross_entropy_loss()")


# ---------------------------------------------------------------------------
# Logistic regression model
# ---------------------------------------------------------------------------


class LogisticRegressionScratch:
    """
    Multinomial logistic regression trained with mini-batch gradient descent.

    This is a LINEAR classifier on top of the embedding features.
    Exactly what sklearn's LogisticRegression does, minus the L2 regularisation.

    Attributes:
      W : weight matrix of shape (D, C)  — D = embed dim, C = num classes
      b : bias vector of shape (C,)
    """

    def __init__(self, n_features: int, n_classes: int, lr: float = 0.1) -> None:
        self.lr = lr
        # Xavier initialisation — keeps activations from exploding/vanishing
        scale = np.sqrt(2.0 / (n_features + n_classes))
        rng = np.random.default_rng(42)
        self.W = rng.normal(0, scale, (n_features, n_classes)).astype(np.float32)
        self.b = np.zeros(n_classes, dtype=np.float32)

    def forward(self, X: np.ndarray) -> np.ndarray:
        """
        Forward pass: return softmax probabilities.

        z = X @ W + b   (shape: N × C)
        p = softmax(z)  (shape: N × C)

        TODO: implement.
        """
        # TODO: implement forward pass
        raise NotImplementedError("TODO: implement forward()")

    def loss(self, X: np.ndarray, y: np.ndarray) -> float:
        """Compute cross-entropy loss on a batch."""
        probs = self.forward(X)
        return cross_entropy_loss(probs, y)

    def gradient_step(self, X: np.ndarray, y: np.ndarray) -> float:
        """
        Compute gradients and update W and b.

        Forward:
          z     = X @ W + b                    shape: (N, C)
          probs = softmax(z)                   shape: (N, C)
          loss  = cross_entropy(probs, y)      scalar

        Backward (the beautiful result of differentiating softmax + cross-entropy):
          dL/dz = probs - one_hot(y)           shape: (N, C)
                  (predicted probs minus the true 1-hot target)

          dL/dW = X.T @ dL/dz  / N            shape: (D, C)
          dL/db = mean(dL/dz, axis=0)          shape: (C,)

        Update:
          W -= lr * dL/dW
          b -= lr * dL/db

        Return the scalar loss (for logging).

        TODO: implement this method — translate the math above into numpy.
          - Run the forward pass (self.forward) to get probs, and record the loss
            (self.loss or cross_entropy_loss) so you can return it at the end.
          - Build the one-hot target matrix for y (shape N×C): a zeros array with a
            single 1 per row in the true-class column.
          - Form dz = probs - one_hot, then turn it into the two gradients using the
            dL/dW and dL/db formulas above (note the 1/N and the `axis=0` mean).
          - Take one gradient-descent step on self.W and self.b (subtract lr * grad),
            and return the scalar loss.
        """
        # TODO: implement gradient step
        raise NotImplementedError("TODO: implement gradient_step()")

    def predict(self, X: np.ndarray) -> np.ndarray:
        """Return the predicted class index for each row of X."""
        return np.argmax(self.forward(X), axis=1)

    def accuracy(self, X: np.ndarray, y: np.ndarray) -> float:
        """Fraction of correct predictions."""
        return float(np.mean(self.predict(X) == y))


# ---------------------------------------------------------------------------
# Training loop
# ---------------------------------------------------------------------------


def train(
    model: LogisticRegressionScratch,
    X_train: np.ndarray,
    y_train: np.ndarray,
    epochs: int = 200,
    batch_size: int = 16,
    print_every: int = 20,
) -> list[float]:
    """
    Mini-batch gradient descent training loop.

    Each epoch:
      1. Shuffle the training set.
      2. Split into mini-batches of `batch_size`.
      3. Call model.gradient_step() on each batch.
      4. Record the mean loss over all batches.

    This function is complete — you don't need to edit it.
    It calls model.gradient_step() which you implement above.
    """
    N = len(y_train)
    loss_history: list[float] = []
    rng = np.random.default_rng(42)

    for epoch in range(epochs):
        perm = rng.permutation(N)
        X_shuf, y_shuf = X_train[perm], y_train[perm]

        epoch_losses: list[float] = []
        for start in range(0, N, batch_size):
            Xb = X_shuf[start : start + batch_size]
            yb = y_shuf[start : start + batch_size]
            loss = model.gradient_step(Xb, yb)
            epoch_losses.append(loss)

        mean_loss = float(np.mean(epoch_losses))
        loss_history.append(mean_loss)

        if (epoch + 1) % print_every == 0 or epoch == 0:
            print(f"  Epoch {epoch + 1:>4}/{epochs}  loss={mean_loss:.4f}")

    return loss_history


# ---------------------------------------------------------------------------
# Harness
# ---------------------------------------------------------------------------


def main() -> None:
    provider = get_provider()
    print(f"\nUsing provider: {provider.name} (embed model: {provider.embed_model})")

    # ── Load & embed ──────────────────────────────────────────────────────
    print("\n[1/4] Loading and embedding dataset...")
    X, y = load_and_embed(provider)
    print(f"  X shape: {X.shape}   (N={X.shape[0]}, D={X.shape[1]})")
    print(f"  y shape: {y.shape}   classes: {NUM_CLASSES}")

    # ── Split ─────────────────────────────────────────────────────────────
    print("\n[2/4] Stratified 80/20 train/test split...")
    X_train, X_test, y_train, y_test = stratified_split(X, y)
    print(f"  Train: {len(y_train)}  Test: {len(y_test)}")

    # ── Build model ───────────────────────────────────────────────────────
    D = X.shape[1]   # embedding dimension
    print(f"\n[3/4] Building LogisticRegressionScratch(D={D}, C={NUM_CLASSES})...")
    model = LogisticRegressionScratch(n_features=D, n_classes=NUM_CLASSES, lr=0.5)

    # ── Train ─────────────────────────────────────────────────────────────
    print("\n[4/4] Training with mini-batch gradient descent...")
    print(f"  Initial train loss: {model.loss(X_train, y_train):.4f}")
    print(f"  Initial train acc:  {model.accuracy(X_train, y_train):.2%}")
    print()

    loss_history = train(model, X_train, y_train, epochs=300, batch_size=16, print_every=50)

    # ── Evaluate ──────────────────────────────────────────────────────────
    train_acc = model.accuracy(X_train, y_train)
    test_acc = model.accuracy(X_test, y_test)

    print(f"\n  Final train loss: {loss_history[-1]:.4f}")
    print(f"  Train accuracy:   {train_acc:.2%}")
    print(f"  Test accuracy:    {test_acc:.2%}")

    # Per-class breakdown
    print("\n  Per-class test accuracy:")
    for cls_idx, cls_name in enumerate(LABEL_NAMES):
        mask = y_test == cls_idx
        if mask.sum() == 0:
            continue
        cls_acc = float(np.mean(model.predict(X_test[mask]) == y_test[mask]))
        print(f"    {cls_name:<12} : {cls_acc:.2%} (n={mask.sum()})")

    print("\n  Compare to Task 3 results — is from-scratch LR close to sklearn LR?")
    print("  Reflection: the gradient formula dL/dz = p - one_hot(y) is the")
    print("  'beautiful' result of differentiating softmax + cross-entropy jointly.")


if __name__ == "__main__":
    main()

File: modules/08-classification/ts/04-logistic-scratch.ts
/**
 * Task 4 🔴 — Multinomial logistic regression FROM SCRATCH (plain arrays, no ML lib).
 *
 * What you'll learn:
 *   - The softmax function: turning raw scores into a probability distribution
 *   - Cross-entropy loss: the standard loss for classification
 *   - Gradient descent: how the weight matrix updates to minimise loss
 *   - Why the gradient of softmax + cross-entropy is elegantly simple
 *
 * The math (README explains each step in plain English):
 *
 *   Forward pass:
 *     z = X @ W + b           (linear layer: batch of scores, shape [N, C])
 *     p = softmax(z)          (probabilities, each row sums to 1)
 *     L = crossEntropy(p, y)  (scalar loss)
 *
 *   Backward pass:
 *     dL/dz = p - oneHot(y)   (elegant: error is predicted probs minus true probs)
 *     dL/dW = Xᵀ @ dL/dz / N
 *     dL/db = mean(dL/dz, axis=0)
 *
 *   Update:
 *     W -= lr * dL/dW
 *     b -= lr * dL/db
 *
 * No external math library. Plain number[][] arrays.
 *
 * How to run:
 *   pnpm tsx modules/08-classification/ts/04-logistic-scratch.ts
 */

import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
import { getProvider } from "@learn-ai/llm-core";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const LABEL_NAMES = ["technology", "science", "business", "sports", "health", "politics"];
const NUM_CLASSES = LABEL_NAMES.length;

// ---------------------------------------------------------------------------
// Dataset loading
// ---------------------------------------------------------------------------

interface DataItem {
  id: number;
  text: string;
  label: string;
}

function loadDataset(): DataItem[] {
  const path = join(__dirname, "../data/texts.json");
  return JSON.parse(readFileSync(path, "utf-8")) as DataItem[];
}

// ---------------------------------------------------------------------------
// Matrix helpers (plain 2-D arrays: number[][])
// ---------------------------------------------------------------------------

/** Create an N×M matrix filled with zeros. */
function zeros(rows: number, cols: number): number[][] {
  return Array.from({ length: rows }, () => new Array(cols).fill(0) as number[]);
}

/** Matrix multiply: A (N×K) @ B (K×M) → C (N×M). */
function matMul(A: number[][], B: number[][]): number[][] {
  const N = A.length, K = A[0].length, M = B[0].length;
  const C = zeros(N, M);
  for (let i = 0; i < N; i++)
    for (let k = 0; k < K; k++)
      for (let j = 0; j < M; j++)
        C[i][j] += A[i][k] * B[k][j];
  return C;
}

/** Transpose: A (N×M) → Aᵀ (M×N). */
function transpose(A: number[][]): number[][] {
  const N = A.length, M = A[0].length;
  const T = zeros(M, N);
  for (let i = 0; i < N; i++)
    for (let j = 0; j < M; j++)
      T[j][i] = A[i][j];
  return T;
}

/** Add bias vector b (length C) to each row of matrix Z (N×C). */
function addBias(Z: number[][], b: number[]): number[][] {
  return Z.map((row) => row.map((v, j) => v + b[j]));
}

// ---------------------------------------------------------------------------
// Softmax and cross-entropy — implement these
// ---------------------------------------------------------------------------

/**
 * Softmax over each row of a 2-D array.
 *
 * softmax(z)_j = exp(z_j) / sum_k(exp(z_k))
 *
 * Numerically stable: subtract the row-max before exp().
 *
 * TODO: implement row-wise. For each row of Z:
 *   - Find the row's max (`Math.max(...row)`) and subtract it from every entry
 *     before exponentiating — that keeps `Math.exp` from overflowing.
 *   - Exponentiate the shifted values, then divide each by the row's total so the
 *     row sums to 1. Return a new matrix of the same shape.
 */
function softmax(Z: number[][]): number[][] {
  // TODO: implement row-wise numerically stable softmax
  throw new Error("TODO: implement softmax()");
}

/**
 * Mean cross-entropy loss over a batch.
 *
 * L = -1/N * sum_i( log( probs[i][y[i]] ) )
 *
 * probs[i][y[i]] is the predicted probability for the TRUE class of sample i.
 * Clip to 1e-12 to avoid log(0).
 *
 * TODO: implement.
 */
function crossEntropyLoss(probs: number[][], y: number[]): number {
  // TODO: implement
  throw new Error("TODO: implement crossEntropyLoss()");
}

// ---------------------------------------------------------------------------
// Logistic regression model
// ---------------------------------------------------------------------------

class LogisticRegressionScratch {
  W: number[][];  // shape: D × C
  b: number[];    // shape: C

  constructor(
    private nFeatures: number,
    private nClasses: number,
    private lr: number = 0.1,
  ) {
    // Xavier initialisation
    const scale = Math.sqrt(2.0 / (nFeatures + nClasses));
    // Seeded pseudo-random with a simple LCG for reproducibility
    let seed = 42;
    const rand = () => {
      seed = (seed * 1664525 + 1013904223) & 0x7fffffff;
      // Box-Muller transform for Gaussian
      const u1 = (seed & 0xffff) / 0x10000 + 1e-10;
      seed = (seed * 1664525 + 1013904223) & 0x7fffffff;
      const u2 = (seed & 0xffff) / 0x10000;
      return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2) * scale;
    };

    this.W = Array.from({ length: nFeatures }, () =>
      Array.from({ length: nClasses }, () => rand()),
    );
    this.b = new Array(nClasses).fill(0) as number[];
  }

  /** Forward pass: compute softmax probabilities. Shape: (N, C). */
  forward(X: number[][]): number[][] {
    /**
     * TODO: compute the linear scores then squash them.
     *   - Multiply X by this.W with `matMul`, add the bias with `addBias`, and pass
     *     the resulting (N, C) score matrix through `softmax`. Return the probs.
     */
    throw new Error("TODO: implement forward()");
  }

  /** Compute cross-entropy loss on a batch. */
  loss(X: number[][], y: number[]): number {
    return crossEntropyLoss(this.forward(X), y);
  }

  /**
   * Compute gradients and update W and b.
   *
   * Forward:
   *   probs = forward(X)                      shape: (N, C)
   *   loss  = crossEntropy(probs, y)          scalar
   *
   * Backward:
   *   dZ    = probs - oneHot(y)               shape: (N, C)
   *   dW    = Xᵀ @ dZ / N                    shape: (D, C)
   *   db    = mean(dZ, axis=0)               shape: (C,)
   *
   * Update:
   *   W[d][c] -= lr * dW[d][c]   for all d, c
   *   b[c]    -= lr * db[c]      for all c
   *
   * Return the loss (for logging).
   *
   * TODO: implement this method — translate the math above into array ops.
   *   - Forward pass (this.forward) to get probs, and compute the loss with
   *     crossEntropyLoss so you can return it at the end.
   *   - Build dZ from the dL/dz formula: start from probs and, for each sample i,
   *     subtract 1 in the true-class column y[i] (that's probs - oneHot(y)).
   *   - Turn dZ into dW and db using the formulas above — dW uses `matMul` on the
   *     transpose of X and is scaled by 1/N; db is the per-column mean of dZ.
   *   - Subtract lr * gradient from this.W (element-wise) and this.b, then return
   *     the loss.
   */
  gradientStep(X: number[][], y: number[]): number {
    // TODO: implement gradient step
    throw new Error("TODO: implement gradientStep()");
  }

  /** Return predicted class index for each row of X. */
  predict(X: number[][]): number[] {
    const probs = this.forward(X);
    return probs.map((row) => row.indexOf(Math.max(...row)));
  }

  /** Fraction of correct predictions. */
  accuracy(X: number[][], y: number[]): number {
    const preds = this.predict(X);
    return preds.filter((p, i) => p === y[i]).length / y.length;
  }
}

// ---------------------------------------------------------------------------
// Stratified train/test split
// ---------------------------------------------------------------------------

function trainTestSplit(
  X: number[][],
  y: number[],
  testFraction = 0.2,
): { XTrain: number[][]; XTest: number[][]; yTrain: number[]; yTest: number[] } {
  const groups: Record<number, number[]> = {};
  for (let i = 0; i < y.length; i++) {
    (groups[y[i]] ??= []).push(i);
  }
  const trainIdx: number[] = [], testIdx: number[] = [];
  for (const indices of Object.values(groups)) {
    const nTest = Math.max(1, Math.round(indices.length * testFraction));
    testIdx.push(...indices.slice(0, nTest));
    trainIdx.push(...indices.slice(nTest));
  }
  return {
    XTrain: trainIdx.map((i) => X[i]),
    XTest: testIdx.map((i) => X[i]),
    yTrain: trainIdx.map((i) => y[i]),
    yTest: testIdx.map((i) => y[i]),
  };
}

// ---------------------------------------------------------------------------
// Training loop (complete — you don't need to edit this)
// ---------------------------------------------------------------------------

function train(
  model: LogisticRegressionScratch,
  XTrain: number[][],
  yTrain: number[],
  epochs = 200,
  batchSize = 16,
  printEvery = 50,
): number[] {
  const N = yTrain.length;
  const lossHistory: number[] = [];

  for (let epoch = 0; epoch < epochs; epoch++) {
    // Shuffle (deterministic using epoch as seed offset)
    const perm = Array.from({ length: N }, (_, i) => i);
    // Simple Fisher-Yates with seeded rand
    let s = 42 + epoch;
    const lcg = () => { s = (s * 1664525 + 1013904223) & 0x7fffffff; return s / 0x7fffffff; };
    for (let i = N - 1; i > 0; i--) {
      const j = Math.floor(lcg() * (i + 1));
      [perm[i], perm[j]] = [perm[j], perm[i]];
    }

    const Xs = perm.map((i) => XTrain[i]);
    const ys = perm.map((i) => yTrain[i]);

    const batchLosses: number[] = [];
    for (let start = 0; start < N; start += batchSize) {
      const Xb = Xs.slice(start, start + batchSize);
      const yb = ys.slice(start, start + batchSize);
      const loss = model.gradientStep(Xb, yb);
      batchLosses.push(loss);
    }

    const meanLoss = batchLosses.reduce((a, b) => a + b, 0) / batchLosses.length;
    lossHistory.push(meanLoss);

    if ((epoch + 1) % printEvery === 0 || epoch === 0) {
      console.log(`  Epoch ${String(epoch + 1).padStart(4)}/${epochs}  loss=${meanLoss.toFixed(4)}`);
    }
  }

  return lossHistory;
}

// ---------------------------------------------------------------------------
// Harness
// ---------------------------------------------------------------------------

async function main() {
  const provider = getProvider();
  console.log(`\nUsing provider: ${provider.name} (embed model: ${provider.embedModel})`);

  // ── Load & embed ──────────────────────────────────────────────────────────
  console.log("\n[1/4] Loading and embedding dataset...");
  const dataset = loadDataset();
  const texts = dataset.map((d) => d.text);
  const labelToIdx = Object.fromEntries(LABEL_NAMES.map((l, i) => [l, i]));
  const y = dataset.map((d) => labelToIdx[d.label]);

  const result = await provider.embed(texts);
  const X = result.vectors;
  console.log(`  X: ${X.length} × ${X[0].length}   y: ${y.length} labels`);

  // ── Train/test split ──────────────────────────────────────────────────────
  console.log("\n[2/4] Stratified 80/20 split...");
  const { XTrain, XTest, yTrain, yTest } = trainTestSplit(X, y);
  console.log(`  Train: ${yTrain.length}  Test: ${yTest.length}`);

  // ── Build model ───────────────────────────────────────────────────────────
  const D = X[0].length;
  console.log(`\n[3/4] Building LogisticRegressionScratch(D=${D}, C=${NUM_CLASSES})...`);
  const model = new LogisticRegressionScratch(D, NUM_CLASSES, 0.5);

  // ── Train ─────────────────────────────────────────────────────────────────
  console.log("\n[4/4] Training with mini-batch gradient descent...");
  console.log(`  Initial train loss: ${model.loss(XTrain, yTrain).toFixed(4)}`);
  console.log(`  Initial train acc:  ${(model.accuracy(XTrain, yTrain) * 100).toFixed(1)}%\n`);

  train(model, XTrain, yTrain, 300, 16, 50);

  // ── Evaluate ──────────────────────────────────────────────────────────────
  const trainAcc = model.accuracy(XTrain, yTrain);
  const testAcc = model.accuracy(XTest, yTest);

  console.log(`\n  Train accuracy : ${(trainAcc * 100).toFixed(1)}%`);
  console.log(`  Test accuracy  : ${(testAcc * 100).toFixed(1)}%`);

  console.log("\n  Per-class test accuracy:");
  const preds = model.predict(XTest);
  for (let c = 0; c < NUM_CLASSES; c++) {
    const idx = yTest.map((v, i) => (v === c ? i : -1)).filter((i) => i >= 0);
    if (idx.length === 0) continue;
    const correct = idx.filter((i) => preds[i] === c).length;
    console.log(`    ${LABEL_NAMES[c].padEnd(12)}: ${(correct / idx.length * 100).toFixed(0)}% (n=${idx.length})`);
  }

  console.log(
    "\n  Reflection: the gradient dZ = probs - oneHot(y) is the beautiful",
    "\n  result of differentiating softmax + cross-entropy jointly.",
    "\n  The model has no hidden layers — it finds a linear decision surface",
    "\n  in the embedding space.",
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
